### PR TITLE
Add release-process.md

### LINKED ISSUE #856: Add doc/release-process.md
On our path towards a 1.0 release it would help to start with a 0.1 release soon. Ideally, we'd do this in a systematic way and first discuss how we want the process to look like initially.

It would make sense for the process to be similar to Bitcoin Core's. Here are some suggestions, based on my understanding of [bitcoin-maintainer-tools/make-tag.py](https://github.com/bitcoin-core/bitcoin-maintainer-tools/blob/a9eaddb7397f149121a4225b4096b04918d20362/make-tag.py) and [bitcoin/doc/release-process.md](https://github.com/bitcoin/bitcoin/blob/8f7b93047581c67f2133cdb8c7845471de66c30f/doc/release-process.md).

We would copy the start of Core's [configure.ac](https://github.com/bitcoin/bitcoin/blob/dca80ffb45fcc8e6eedb6dc481d500dedab4248b/configure.ac#L10) into ours:
```
define(_CLIENT_VERSION_MAJOR, 0)
define(_CLIENT_VERSION_MINOR, 99)
define(_CLIENT_VERSION_BUILD, 0)
define(_CLIENT_VERSION_RC, 0)
define(_CLIENT_VERSION_IS_RELEASE, false)
```
and replace our `AC_INIT` with
```
AC_INIT([libsecp256k1],m4_join([.], _CLIENT_VERSION_MAJOR, _CLIENT_VERSION_MINOR, _CLIENT_VERSION_BUILD)m4_if(_CLIENT_VERSION_RC, [0], [], [rc]_CLIENT_VERSION_RC))
```
We can print the version at the end of the configure script and a warning if `RELEASE` is not true.

Now in order to create a release candidate we

1. Start [writing release notes](https://github.com/bitcoin/bitcoin/blob/master/doc/release-process.md#write-the-release-notes)
2. Create a corresponding release branch
    1. Create a commit to fix the `VERSIONS` in configure.ac (including `RC=1`)
    2. Tag the commit with a version of make-tag.py that's adapted to libsecp
    3. Create github release
3. To create a new release candidate repeat steps 2.i through 2.iii
4. Once we're happy with the release candidate
    1. Archive the release notes for the new version to doc/release-notes/ (branch master and branch of the release)
    2. Create commit to set `IS_RELEASE` to `true` and `VERSION_RC` to `0`
    3. Tag the release
    4. Create a new GitHub release with a link to the archived release notes

This doesn't look as simple as I would have hoped, but it's not terrible. What do you think?

### PR DESCRIPTION
**Pull request by jonasnick**
This is an attempt at a simple release process. Fixes #856. To keep it simple, there is no concept of release candidates for now.

The release version is determined by semantic versioning of the API. Since it does not seem to be a lot of work, it is proper to also version the ABI with the libtool versioning system. This versioning scheme (semver API, libtool versioning ABI) follows the suggestion in the [autotools mythbusters](https://autotools.io/libtool/version.html).

Experimental modules are a bit of a headache, as expected. This release process suggests to treat any change in experimental modules as backwards compatible. That way, users of stable modules are not bothered by frequent non-backwards compatible releases. But a downside is that one must not use experimental modules in shared libraries (which should be mentioned in the README?). It would be nice if we could make the schnorrsig module stable in the not too distant future (see also #817).

### CODE REVIEW

**Comment by rustyrussell** [comment]
This looks good!  Note that c-lightning has a tool which extracts Changelog entries from commit messages, and also a bot that tests they exist for each PR (or Changelog-None is in GitHub description of PR).

**Comment by prusnak** [comment]
I suggest to use the format described here: https://keepachangelog.com/ - the guide also mentions that ChangeLog is not a git log dump :-)

**Comment by rustyrussell** [comment]
> I suggest to use the format described here: https://keepachangelog.com/ - the guide also mentions that ChangeLog is not a git log dump :-)

Just to be clear: our tool extracts lives starting with Changelog-(Added|Fixed|Changed|Deprecated|Removed) and adds links to the PRs and sorts them :)

**Comment by prusnak** [comment]
> our tool extracts lives

God, what kind of tools do you use?! 😁

**Comment by jonasnick** [comment]
Thanks for the feedback. I added an `Unreleased` section to the changelog and categories as suggested by keepachangelog.com. I think in most cases the titles of PRs are sufficiently descriptive such that CHANGELOG entries can be title of PR + link.

File: configure.ac [hunk 1]
```diff
@@ -1,5 +1,15 @@
...
+
+# The package (a.k.a. release) version is based on semantic versioning 2.0.0 of
+# the API. All changes in experimental modules are treated as
+# backwards-compatible and therefore at most increase the minor version.
+define(_PKG_VERSION_MAJOR, 0)
+define(_PKG_VERSION_MINOR, 1)
+define(_PKG_VERSION_BUILD, 0)
+define(_PKG_VERSION_IS_RELEASE, false)
+
+AC_INIT([libsecp256k1],m4_join([.], _PKG_VERSION_MAJOR, _PKG_VERSION_MINOR, _PKG_VERSION_BUILD)m4_if(_PKG_VERSION_IS_RELEASE, [true], [], [-pre]),[https://github.com/bitcoin-core/secp256k1/issues],[secp256k1],[https://github.com/bitcoin-core/secp256k1])
```
[the remark this answers is not in the document]
Comment (jonasnick) [reply configure.ac]: good point, fixed

**Comment by sipa** [comment]
ACK procedure/changes, but somehow CI does not like this.

**Review by jonasnick — COMMENTED** [review]
rebased

**Review by real-or-random — APPROVED** [review]
utACK 86d0b9eba0b191319c8b6dbfd9d014420f06c836

**Comment by sipa** [comment]
Hmm, what is the expected difference between the package version and the library version? Why would they not be the same?

**Comment by real-or-random** [comment]
@sipa See https://autotools.io/libtool/version.html

@jonasnick Sorry, needs trivial rebase again.

**Comment by sipa** [comment]
Can we write in the release process document how each of the 6 numbers are supposed to be updated as a function of the changes in such releases?

**Comment by sipa** [comment]
Another question (not necessarily something to be explicitly explained in the release process here, but still something that isn't currently clear to me): what branches structure are the release branches going to use? Possibilities:
* All release branches branch off master, and thus there is no (clean) way to create a `x.y.(z+1)` after `x.(y+1).z` has been released.
* The branch for `x.(y+1).0` branches off the `x.y.0` branch; the branch for `x.y.(z+1)` branches off the `x.y.z` one. Open questions with this:
  * How to obtain monotonically increasing library "revision" value in the presence of branches?
  * When creating a new branch by forking an already-released branch, would that imply resetting the IS_RELEASE variable? Or do we create the fork at the last commit before the one that sets the IS_RELEASE variable?

**Review by jonasnick — COMMENTED** [review]
@sipa The rules are described in configure.ac ("The library version is based on libtool versioning of the ABI. The set of rules for updating the version can be found here: https://www.gnu.org/software/libtool/manual/html_node/Updating-version-info.html") or do you mean copying over the rules to the release-process.md? I thought it's best to have the rules where they version is set because it serves both as a doc of the existing values as well as how to update.

> Another question

That's a good question. I'd like it answered in the release_process.md. How does Bitcoin Core do this? I would have thought everything is based off master except patches to earlier versions are done by branching off the earlier release branch. I think this would only have to be done for patches because it's unlikely we release a x.(y+1).z when (x+1).y.z is already released. Btw, "revision" isn't monotonically increasing, it's set to if there's any change to the interface (including additions).

**Comment by sipa** [comment]
@jonasnick Yes, I meant including the actual rules for incrementing the library version somewhere, not just a reference. I find the rules described in that link kind of vague, and I also had never heard about the concept of library versions at all; if someone is in a similar situation, and just sees the link they may not realize it's actually explaining why it's different from the numbers above.

> That's a good question. I'd like it answered in the release_process.md. How does Bitcoin Core do this? I would have thought everything is based off master except patches to earlier versions are done by branching off the earlier release branch. I think this would only have to be done for patches because it's unlikely we release a x.(y+1).z when (x+1).y.z is already released. Btw, "revision" isn't monotonically increasing, it's set to if there's any change to the interface (including additions).

Bitcoin Core's release process is roughly:
* As a general rule, with very few exceptions, all PRs are made to the master branch.
* Every so often the master branch is forked off to create a major release branch (e.g. the [`22.x`](https://github.com/bitcoin/bitcoin/tree/22.x) branch). After that point, future changes in master *may* be backported to the `22.x` branch if they're bugfixes or somehow specifically relevant for that release. At some point, release candidates `v22.0rc1`, `v22.0rc2`, ... are created from this branch, and finally the final `v22.0` tag is created.
* Minor releases (e.g. `v22.1`) are created by continuing to backport changes (bugfixes, mostly) from the master branch into the `22.x` branch after `v22.0` is released, en then later in the same branch tagging `v22.1rc1`, `v22.1rc2`, ..., and `v22.1`. These can be made for multiple major release branches simultnaneously (e.g. a bugfix in master is sufficiently important that it deserves going into major releases 21 and 20 too, could be backported to all these branches, and then trigger release candidates/releases of new minor releases in all of them.

So the major difference with what you're proposing is that Bitcoin Core only has branches for major releases, not for very release, and that branch is then used for potentially multiple tags.

**Comment by sipa** [comment]
@jonasnick Oh, I noticed that the link in the source is different from the one @real-or-random linked to. That one gives much more concrete instructions already.

**Comment by jonasnick** [comment]
I don't see signficant down or upsides with either branching approach so I'd follow Bitcoin Core's process where applicable because that was the initial intention anyway.

> It's set to what if there is any change to the interface?

Lol, sorry, I have a new (terrible) keyboard. 0.

**Comment by sipa** [comment]
Well, I think there are probably some differences that could justify deviating. For one, I don't think we expect many "major" releases (in the semver sense) after 1.0, certainly not two per year.

**Comment by jonasnick** [comment]
Yes, I would have translated Core major -> secp minor and Core minor -> patch. So we'd make one branch per minor release and tag patch versions.

File: doc/release-process.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# Release Process
+
+1. Open PR to master that
+   1. adds release notes to `doc/CHANGELOG.md` and
+   2. if this is **not** a patch release, updates `_PKG_VERSION_{MAJOR,MINOR}` and `_LIB_VERSIONS_*` in `configure.ac`
+2. After the PR is merged,
+   * if this is **not** a patch release, create a release branch with name `MAJOR.MINOR.x`.
```
Comment (real-or-random) [inline doc/release-process.md:7]: nit: So I assume `x` is a literal x here, right, so this is `1.2.x`. Then we could just call it `1.2`. I think this is in fact closer to Core (just `MAJOR.MINOR`), and this is how you usually abbreviate versions: Just drop the parts you don't care about.

**Review by real-or-random — APPROVED** [review]
I agree with all the additions, this looks good.

File: doc/release-process.md [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# Release Process
+
+1. Open PR to master that
+   1. adds release notes to `doc/CHANGELOG.md` and
```
Comment (t-bast) [inline doc/release-process.md:4]: In several projects, I've found it quite painful to write all the release notes at once when we actually do a release (the burden of writing the release notes falls on the guy who publishes the release and needs to document other people's additions to the codebase).

Something that worked much better for me was to incrementally update the future release notes: each PR to master that will impact the next release should update the future release notes. This way when we want to publish a release, we only need to do a bit of clean-up, but the release notes are already written and up-to-date. It needs a bit of effort at first to remind contributors to do it, but in the end we usually end up with better release notes.

Feel free to ignore if you think that doesn't make sense for this project, but I felt it was worth sharing in case it can be helpful.

Comment (real-or-random) [reply doc/release-process.md:4]: I think in principle I agree but I'm not really sure. My experience in this repo is that it happens quite often that you ask a  contributor to fix a nit before you ACK and then the PR is stalled for months. ;)

Comment (jonasnick) [reply doc/release-process.md:4]: Agree on @real-or-random's point but we can still remind regular contributors to do this. I wouldn't want to force this on irregular contributors because it's additional work that discourages from contributing PRs in the first place. We could add a pull request template that optionally requests a changelog update.

One significant downside of this is that pull requests would all edit similar space in the same file which, when one PR is merged, would require rebasing every other PR.

File: doc/release-process.md (hunk 2 above)
Comment (jonasnick) [reply doc/release-process.md:7]: Core creates a MAJOR.x branch (where x is a literal x) according to @sipa's description ([core branches](https://github.com/bitcoin/bitcoin/branches)).

File: doc/release-process.md (hunk 3 above)
Comment (real-or-random) [reply doc/release-process.md:4]: > One significant downside of this is that pull requests would all edit similar space in the same file which, when one PR is merged, would require rebasing every other PR.

Apparently that's a real problem which led to a "crisis" in the GitLab project and needs an entirely non-trivial solution:
https://about.gitlab.com/blog/2018/07/03/solving-gitlabs-changelog-conflict-crisis/

I think then the maintainers should create the changelog... I suppose it will be easier for everyone (including maintainers).

Comment (sipa) [reply doc/release-process.md:4]: Bitcoin Core uses `doc/release-notes-<PRNUMBER>.md` files which are added by individual PRs, and are then merged into a release release-notes.md by maintainers.

File: doc/release-process.md (hunk 2 above)
Comment (sipa) [reply doc/release-process.md:7]: I'm fine with either. I think Bitcoin Core went with "22.x" because branches that are just a number are a bit weird, but in the end it's all just arbitrary conventions.

**Comment by sipa** [comment]
LGTM, feel free to squash.

File: doc/release-process.md (hunk 2 above)
Comment (jonasnick) [reply doc/release-process.md]: Ok, I removed the .x because it looks nicer.

File: doc/release-process.md (hunk 3 above)
Comment (real-or-random) [reply doc/release-process.md:4]: Just for reference, here's some docs on the process:
https://github.com/bitcoin/bitcoin/blob/23ae7931be50376fa6bda692c641a3d2538556ee/doc/release-process.md#write-the-release-notes
https://github.com/bitcoin/bitcoin/blob/23ae7931be50376fa6bda692c641a3d2538556ee/doc/release-notes.md

This has some helpful notes, e.g., a script to generate a (draft) changelog from git history. This may be helpful to us.

**Review by real-or-random — APPROVED** [review]
ACK c8fd83fb834321fcadd062709624f6fb9a95a40d

**Comment by jonasnick** [comment]
rebased

**Comment by apoelstra** [comment]
utACK 3ed0d02bf7690c68ba3ba74da765614809352049

Looks great!

**Review by real-or-random — APPROVED** [review]
ACK 3ed0d02bf7690c68ba3ba74da765614809352049

**Comment by elichai** [comment]
ACK b4b02fd8c4276779f115af5985508f54ef9e51be

**Comment by real-or-random** [comment]
@elichai wrong commit :P

**Comment by elichai** [comment]
@real-or-random Ops, somehow I keep getting confused on the order of commits in github (in git the latest is on top, in github it's at the bottom)

ACK 3ed0d02bf7690c68ba3ba74da765614809352049

**Comment by apoelstra** [comment]
@elichai if you have
```
[remote "origin"]
        fetch = +refs/pull/*:refs/remotes/pr/*
```
in your gitconfig, you can do something like `git rev-parse pr/964/head | xsel` to get the top commit ID.

You may still type the wrong PR number or write "ACK" in the wrong browser tab, but at least there's one less mistake you can make :)

**Comment by sipa** [comment]
ACK 3ed0d02bf7690c68ba3ba74da765614809352049
